src/Translator/assets/src/types.d.ts

```diff
@@ -0,0 +1,28 @@
+type DomainType = string;
+type LocaleType = string;
+type TranslationsType  = Record<DomainType, { parameters: ParametersType }>;
+type NoParametersType = Record<string, never>;
+type ParametersType = Record<string, string | number> | NoParametersType;
+type RemoveIntlIcuSuffix<T extends string> = T extends `${infer U}+intl-icu` ? U : T;
+
+export type DomainsOf<M> = M extends Message<infer Translations, LocaleType> ? RemoveIntlIcuSuffix<keyof Translations>: never;
+export type LocaleOf<M> = M extends Message<TranslationsType, infer Locale> ? Locale : never;
+export type ParametersOf<M, D extends DomainType> = M extends Message<infer Translations, LocaleType>
+    ? (Translations[D] extends { parameters: infer Parameters }
+        ? Parameters
+        : never)
+    : never;
```

Comment:
These types make my head spin a bit... did you write these by hand?

Reply:
Yep, this was written by hand! This can be a bit complicated at first, but this is easier if you understaind `infer`, `extends`, and `never` keywords.

For the given type: 
```ts
export type ParametersOf<M, D extends DomainType> = M extends Message<infer Translations, LocaleType>
    ? Translations[D] extends { parameters: infer Parameters }
        ? Parameters
        : never
    : never;
```
1. We define a type `ParametersOf<M, D extends DomainType>`, which use two generics:
    1. `M` (a `Message` or anything else) 
    2. and `D extends DomainType` (a translation domain)
2. If `M` is a `Message`: 
    1. then we fetch the translations, thanks to `infer Translations`, 
    2. otherwise we stop, thanks to `never`
3. If `Translations[D]` (translations for the domain `D`) is a JS object containing one key `parameters`:
    1. then we returns the `Parameters`'s type, thanks to `infer Parameters`
    2. otherwise we stop, thanks to `never`